kaggle-classification/trainer/model.py

```diff
@@ -144,56 +142,68 @@ def estimator_spec_for_softmax_classification(logits, labels, mode):
-        strides=CNN_PARAMS.POOLING_STRIDE,
-        padding='SAME')
-    # Transpose matrix so that n_filters from convolution becomes width.
-    pool1 = tf.transpose(pool1, [0, 1, 3, 2])
-  with tf.variable_scope('CNN_Layer2'):
-    # Second level of convolution filtering.
-    conv2 = tf.layers.conv2d(
-        pool1,
-        filters=CNN_PARAMS.N_FILTERS,
-        kernel_size=CNN_PARAMS.FILTER_SHAPE2,
-        padding='VALID')
-    # Max across each filter to get useful features for classification.
-    pool2 = tf.squeeze(tf.reduce_max(conv2, 1), squeeze_dims=[1])
-
-  # Apply regular WX + B and classification.
-  logits = tf.layers.dense(pool2, MAX_LABEL, activation=None)
-  predicted_classes = tf.argmax(logits, 1)
+  embedding_size = CNN_PARAMS.EMBEDDING_SIZE
+  filter_sizes = CNN_PARAMS.FILTER_SIZES
+  num_filters = CNN_PARAMS.N_FILTERS
+  dropout_keep_prob = CNN_PARAMS.DROPOUT_KEEP_PROB
+
+  with tf.name_scope("embedding"):
+    W = tf.Variable(
+        tf.random_uniform([n_words, embedding_size], -1.0, 1.0),
+        name="W")
+
+    embedded_chars = tf.nn.embedding_lookup(W, features[WORDS_FEATURE])
+    embedded_chars_expanded = tf.expand_dims(embedded_chars, -1)
+
+  pooled_outputs = []
+  for i, filter_size in enumerate(filter_sizes):
+    with tf.name_scope("conv-maxpool-%s" % filter_size):
+
+      # Convolution Layer
+        filter_shape = [filter_size, embedding_size, 1, num_filters]
+        W = tf.Variable(tf.truncated_normal(filter_shape, stddev=0.1), name="W")
+        b = tf.Variable(tf.constant(0.1, shape=[num_filters]), name="b")
+        conv = tf.nn.conv2d(
+          embedded_chars_expanded,
+          W,
+          strides=[1, 1, 1, 1],
+          padding="VALID",
+          name="conv")
+        # Apply nonlinearity
+        hh = tf.nn.relu(tf.nn.bias_add(conv, b), name="relu")
+
+        # Max-pooling over the outputs. Max over samples in batch and
+        # all filters.
+        pooled = tf.nn.max_pool(
+          hh,
+          ksize=[1, MAX_DOCUMENT_LENGTH - filter_size + 1, 1, 1],
+          strides=[1, 1, 1, 1],
+          padding='VALID',
+          name="pool")
+
+        pooled_outputs.append(pooled)
+
+  # Combine all the pooled features
+  num_filters_total = num_filters * len(filter_sizes)
+  h_pool = tf.concat(pooled_outputs, 3)
+  h_pool_flat = tf.reshape(h_pool, [-1, num_filters_total])
+
+  # Add dropout in training
+  with tf.name_scope("dropout"):
+    # Set dropout rate to 1 (disable dropout) by default
+    h_drop = tf.nn.dropout(h_pool_flat, 1.0)
+
+    if mode == tf.estimator.ModeKeys.TRAIN:
+      h_drop = tf.nn.dropout(h_pool_flat, dropout_keep_prob)
+
+  # Add a fully connected layer to do prediction
+  with tf.name_scope("output"):
+    W = tf.Variable(tf.truncated_normal([num_filters_total, MAX_LABEL], stddev=0.1), name="W")
+    b = tf.Variable(tf.constant(0.1, shape=[MAX_LABEL]), name="b")
+    scores = tf.nn.xw_plus_b(h_drop, W, b, name="scores")
```

Comment:
This works. You can do it in fewer lines of code with tf.layers.dense.

Reply:
Ok, good to know!